File: frontend/app/contextTelemetry.ts
export type ContextCategory = "system" | "tools" | "media" | "history" | "memory" | "evidence";

export type ContextCall = {
  call_index: number;
  model?: string | null;
  context_role?: string | null;
  declared_max?: number | null;
  effective_limit: number;
  prompt_budget: number;
  estimated_prompt_tokens: number;
  actual_prompt_tokens?: number | null;
  actual_completion_tokens?: number | null;
  completion_reserve: number;
  measurement: "exact" | "estimated";
  categories: Record<ContextCategory, number>;
  pressure: "normal" | "trim_tools" | "summarize_history" | "essential_only";
  compacted: boolean;
  overflow_retry: boolean;
  changes: {
    history: { summarized: number; dropped: number };
    evidence?: { summarized: number; dropped: number };
    tools: { trimmed: number };
  };
};

export type ContextTelemetry = {
  version: number;
  calls: ContextCall[];
  primary_call: number;
  final_call: number;
  aggregation: "per_call_not_summed";
  compacted: boolean;
  overflow_retried: boolean;
};

export function primaryContextCall(report?: ContextTelemetry): ContextCall | null {
  if (report?.version !== 1 || !Array.isArray(report.calls) || !report.calls.length) return null;
  return report.calls[report.primary_call] ?? report.calls.at(-1) ?? null;
}

export function affectedContextCalls(report?: ContextTelemetry): ContextCall[] {
  if (report?.version !== 1 || !Array.isArray(report.calls)) return [];
  return report.calls.filter(
    (call, index) => index !== report.primary_call && (call.compacted || call.overflow_retry),
  );
}

export function contextMeter(report?: ContextTelemetry) {
  const call = primaryContextCall(report);
  if (!call || !Number.isFinite(call.effective_limit) || call.effective_limit <= 0) {
    return { state: "missing" as const, call: null, used: 0, denominator: 0, percent: 0 };
  }
  const used = call.actual_prompt_tokens ?? call.estimated_prompt_tokens;
  const state = call.overflow_retry
    ? "retried" as const
    : call.compacted
      ? "compacted" as const
      : call.pressure !== "normal"
        ? "near_limit" as const
        : call.measurement === "estimated" ? "estimated" as const : "normal" as const;
  return {
    state,
    call,
    used,
    denominator: call.effective_limit,
    percent: Math.min(100, Math.round((used / call.effective_limit) * 100)),
  };
}

File: backend/agents/context_telemetry.py
"""Safe, versioned context telemetry for WebSocket clients and persistence."""

from __future__ import annotations

from agents.context_manager import ContextReport


_CATEGORIES = ("system", "tools", "media", "history", "memory", "evidence")


def _public_call(report: ContextReport, index: int) -> dict:
    """Project a report onto an explicit allowlist of numeric/status fields."""
    return {
        "call_index": index,
        "model": report.model,
        "context_role": report.context_role,
        "declared_max": report.declared_context,
        "effective_limit": report.context_window,
        "prompt_budget": report.prompt_budget,
        "estimated_prompt_tokens": report.estimated_prompt_tokens,
        "actual_prompt_tokens": report.actual_prompt_tokens,
        "actual_completion_tokens": report.actual_completion_tokens,
        "completion_reserve": report.completion_reserve,
        "measurement": "exact" if report.actual_prompt_tokens is not None else "estimated",
        "categories": {name: int(report.categories.get(name, 0)) for name in _CATEGORIES},
        "pressure": report.pressure,
        "compacted": report.compacted,
        "overflow_retry": report.retry,
        "changes": {
            "history": {
                "summarized": report.summarized_categories.get("history", 0),
                "dropped": report.removed_categories.get("history", 0),
            },
            "evidence": {
                "summarized": report.summarized_categories.get("evidence", 0),
                "dropped": report.removed_categories.get("evidence", 0),
            },
            "tools": {"trimmed": report.trimmed_tool_messages},
        },
    }


def build_context_telemetry(reports: list[ContextReport]) -> dict | None:
    """Return per-call telemetry; never sum incompatible request windows."""
    if not reports:
        return None
    calls = [_public_call(report, index) for index, report in enumerate(reports)]
    final_index = len(calls) - 1
    return {
        "version": 1,
        "calls": calls,
        # Different provider calls can have different effective windows. Keep
        # them inspectable and display the final call instead of summing them.
        "primary_call": final_index,
        "final_call": final_index,
        "aggregation": "per_call_not_summed",
        "compacted": any(call["compacted"] for call in calls),
        "overflow_retried": any(call["overflow_retry"] for call in calls),
    }
